### PR TITLE
feat:  grepable output

### LINKED ISSUE #57: Grepable output
✔, ✘

### PR DESCRIPTION
**Pull request by DnFreddie**
Introduced grepable formatting and added a Format struct to the Report for better organization and extensibility. Simplified the main function by decluttering unnecessary variables and centralizing output handling.

Fixes #57

### CODE REVIEW

File: internal/report.go [hunk 1]
```diff
@@ -17,5 +31,66 @@ type Report struct {
 	// Network error.
 	Error error `json:"error,omitempty"`
 	// Extra information. Depends on the protocol.
-	Extra string `json:"extra,omitempty"`
+	Extra  string `json:"extra,omitempty"`
+	Format Format
+}
+
+func (r *Report) PrintFormatted(f Format) {
+
```
Comment (jesusprubio) [inline internal/report.go:39]: Please, remove these blank lines inside of the function definitions.

File: internal/report.go [hunk 2]
```diff
@@ -17,5 +31,66 @@ type Report struct {
...
+	case HumanFormat:
+		printHumanFormat(r)
+	case JSONFormat:
+		printJSONFormat(r)
+	case GrepFormat:
+		printGrepableFormat(r)
+	}
+
+}
+func printJSONFormat(r *Report) {
```
Comment (jesusprubio) [inline internal/report.go:50]: I would made these part of the `Report` definition: `func (r *Report) printJSONFormat() {`

File: internal/report.go [hunk 3]
```diff
@@ -17,5 +31,66 @@ type Report struct {
...
+}
+
+var (
+	green = color.New(color.FgGreen).SprintFunc()
+	red   = color.New(color.FgRed).SprintFunc()
+	bold  = color.New(color.Bold).SprintFunc()
+	faint = color.New(color.Faint).SprintFunc()
+)
+
+func printHumanFormat(r *Report) {
```
Comment (jesusprubio) [inline internal/report.go:66]: Same, these can be part of the `Report` definition.

File: internal/report.go [hunk 4]
```diff
@@ -17,5 +31,66 @@ type Report struct {
...
+		prefix = red("✘")
+		suffix = r.Error.Error()
+	}
+	suffix = fmt.Sprintf("(%s)", suffix)
+
+	fmt.Printf("%s %s %s\n", prefix, line, faint(suffix))
+}
+
+// Output: HTTP/1.1    2024-11-18T15:00:00Z    192.168.1.1    success    Request processed successfully
+func printGrepableFormat(r *Report) {
```
Comment (jesusprubio) [inline internal/report.go:82]: Same.

File: internal/report.go [hunk 5]
```diff
@@ -17,5 +31,66 @@ type Report struct {
 	// Network error.
 	Error error `json:"error,omitempty"`
 	// Extra information. Depends on the protocol.
-	Extra string `json:"extra,omitempty"`
+	Extra  string `json:"extra,omitempty"`
+	Format Format
+}
+
+func (r *Report) PrintFormatted(f Format) {
```
Comment (jesusprubio) [inline internal/report.go:38]: I think this one (and the ones called here)  should return an `string`, printing it only in the main function.

File: internal/report.go (hunk 2 above)
Comment (jesusprubio) [reply internal/report.go:50]: Like before, this one should also return a `string`.

File: main.go [hunk 6]
```diff
@@ -104,19 +103,18 @@ func main() {
 	}
 	go func() {
 		logger.Debug("Listening for reports ...")
+		// Print Report Lines
```
Comment (jesusprubio) [inline main.go:106]: I think we don't need this comment. If you think we do please change it to 'Print report lines.`

File: main.go [hunk 7]
```diff
@@ -104,19 +103,18 @@ func main() {
...
 		for report := range probe.ReportCh {
 			logger.Debug("New report", "report", *report)
-			var line string
+
+			format := internal.HumanFormat // Default format
```
Comment (jesusprubio) [inline main.go:110]: This line is also clear enough without the comment.

File: main.go [hunk 8]
```diff
@@ -104,19 +103,18 @@ func main() {
...
 			if opts.JSONOutput {
```
Comment (jesusprubio) [inline main.go:111]: I would use a `switch` here.

File: internal/report.go (hunk 2 above)
Comment (jesusprubio) [reply internal/report.go:50]: So I would also rename these to: `NewLine()`, `newLineJSON`, `newLineGrep`, etc.

File: main.go [hunk 9]
```diff
@@ -104,19 +103,18 @@ func main() {
...
-				line = string(reportJSON)
-			} else {
-				line = reportToLine(report)
+				format = internal.JSONFormat
+			} else if opts.GrepFormat {
+				format = internal.GrepFormat
 			}
-			fmt.Println(line)
+			report.PrintFormatted(format)
+
```
Comment (jesusprubio) [inline main.go:117]: Please, remember to remove all blank lines inside functions.

**Review by jesusprubio — CHANGES_REQUESTED** [review]
Thank you! I appreciate it.

Please remember to run this before finishing the PR: https://github.com/jesusprubio/up/blob/main/CONTRIBUTING.md#test

File: internal/report.go [hunk 10]
```diff
@@ -17,5 +31,66 @@ type Report struct {
...
+	case GrepFormat:
+		printGrepableFormat(r)
+	}
+
+}
+func printJSONFormat(r *Report) {
+
+	reportJSON, err := json.Marshal(r)
+	if err != nil {
+		log.Fatal(fmt.Errorf("marshaling report: %w", err))
```
Comment (jesusprubio) [inline internal/report.go:54]: Should return the error. A library shouldn't log anything.

**Comment by DnFreddie** [comment]
I resolved all the comments u gave me this now should be fine to merge

**Pull request merged**